Update local cluster setup & documentation (#2494)
elastisys/compliantkubernetes-apps: DEVELOPMENT.md

This requires that `kind` is installed and that either `podman` or `docker` is available and setup to work with `kind`.

### Terminology

> - `apps-flavor` - one of `prod`, `dev` or `baremetal` is a global switch for configuring the clusters and deployed applications in different ways. The recommended value when working with local clusters is `dev`.

Reviewer: ```suggestion
- `apps-flavor` - one of `prod`, `dev` or `air-gapped` is a global switch for configuring the clusters and deployed applications in different ways. The recommended value when working with local clusters is `dev`.
```

> - `local-cluster-profile` - a `Cluster` configuration object passed to `kind` when creating or updating local clusters. We use profiles to differentiate between single-/multi-node clusters and/or to enable special features such as container image caching. Use `./bin/local-clusters.sh list profiles` to see a list of built-in profiles.

Reviewer: To be pedantic (as it is written `<kind-config|local-cluster-profile>` in usage latter), the profiles as the preconfigured `Cluster` configs.

> - `domain` - a local domain name. Using a subdomain of `.dev-ck8s.com` allows for setting up DNS based challenges for certificates issued by `cert-manager`, thus avoiding the use of self-signed certificates.

Reviewer: The part about `.dev-ck8s.com` is internal, so that should not be included.

> sysctl net.ipv4.ip_unprivileged_port_start # expected at 53
> ```

> [!tip]
> Since local clusters are effectively ephemeral they can pull a lot of images and `kind` has no built-in system to manage images.
> So, for the local clusters script there are commands to create and delete local pull through registry caches for a few upstream registries.
> The command to do so is `./scripts/local-cluster.sh cache <create|delete>`, then one can make use of the local cluster profiles `<single|multi>-node-cache` that are prepared to use it by default.

As a first step, configure the clusters:

```sh
# with CK8S_CONFIG_PATH and CK8S_PGP_FP set
./scripts/local-cluster.sh config <name> <apps-flavor> <domain>
```

This will use the default local listen address of `127.0.64.43` for the SC worker node and `127.0.64.143` for the WC worker node.

Apps will be configured with `ck8sCloudProvider: none` and `ck8sFlavor: <apps-flavor>` and set some default values to run on local clusters.

> By default, local clusters will use `calico` for networking, `local-path-provisioner` for block storage, and `minio` for object storage.

Reviewer: It feels like the row "...default local listen address..." is a bit misplaced? :thinking:

Author: I've rephrased it a bit and hoisted the "Apps will be configured.." paragraph on top. It is a bit weirdly placed, but I think it's important to communicate that we're configuring the two clusters with different listen addresses.

We'll also be passing the `--skip-minio` options for the WC, to deploy block storage only to the SC cluster.

```sh
./scripts/local-cluster.sh create <name>-sc <kind-config-file|local-cluster-profile>
./scripts/local-cluster.sh create <name>-wc <kind-config-file|local-cluster-profile> --skip-minio
```

> [!important]
> To support the integrated SC <--> WC scenario we require that the cluster names be suffixed with either `-sc` or `-wc`

#### In-cluster DNS setup

Making the clusters "aware" of each other will require custom DNS configuration snippets to both the SC and WC that ensure:

> - `(grafana|harbor|opensearch|ops).<domain>` resolve to an SC worker node IP from within the WC cluster

Reviewer 2: These subdomains are configurable, not sure if that should be handled by this or if it's enough leave a note that it's not supported?

> - `(grafana|harbor|opensearch|ops).<domain>` resolve to an SC worker node IP from within the WC cluster
> - all other `.<domain>` queries from within the WC cluster resolve to the service IP of the `nginx-ingress`
> - `(grafana|harbor|opensearch|ops).<domain>` resolve to the service IP of the `nginx-ingress` within the SC cluster
> - all other `.<domain>` queries from within the SC cluster resolve to a WC worker node IP

Reviewer: Suggestion to make it clearer, nit on `nginx-ingress` -> `ingress-nginx`.
```suggestion
- From within the WC:
    - resolve`(grafana|harbor|opensearch|ops).<domain>` queries to a SC worker node IP.
    - resolve all other `.<domain>` queries to the service IP of `ingress-nginx` in the WC.
- From within the SC:
     - resolve `(grafana|harbor|opensearch|ops).<domain>` queries to the service IP of the `ingress-nginx` in the SC.
     - resolve all other `.<domain>` queries to a WC worker node IP.
```

Since these snippets depend on the IP addresses of running containers, they must be setup _after_ the initial `create` commands.

The following command will configure node-local DNS and (re)deploy the `node-local-dns` stack:

```sh
./scripts/local-cluster-sh setup_node_local_dns
```

#### Self-signed certificate setup

> Add the following block to `$CK8S_CONFIG_PATH/common-config.yaml` (you might need to merge the keys in `global` map manually):

Reviewer: Suggestion, very much optional:

You can add some additional thing here, that "if you cannot setup cert-manager to use DNS-01 challenges you have the option of using self-signed certificates" or something.

> sysctl net.ipv4.ip_unprivileged_port_start # expected at 53
> ```

> [!tip]
> Since local clusters are effectively ephemeral they can pull a lot of images and `kind` has no built-in system to manage images.
> So, for the local clusters script there are commands to create and delete local pull through registry caches for a few upstream registries.
> The command to do so is `./scripts/local-cluster.sh cache <create|delete>`, then one can make use of the local cluster profiles `<single|multi>-node-cache` that are prepared to use it by default.

As a first step, configure the clusters:

```sh
# with CK8S_CONFIG_PATH and CK8S_PGP_FP set
./scripts/local-cluster.sh config <name> <apps-flavor> <domain>
```

Apps will be configured with `ck8sCloudProvider: none` and `ck8sFlavor: <apps-flavor>` and set some default values to run on local clusters.
By default, local clusters will use `calico` for networking, `local-path-provisioner` for block storage, and `minio` for object storage.

The local listen address will be set to `127.0.64.43` for the SC worker node, and `127.0.64.143` for the WC worker node.

> We'll also be passing the `--skip-minio` options for the WC, to deploy block storage only to the SC cluster.

Author: Good catch! minio provides object (S3-like) storage, not block

helmfile -e workload_cluster <operation> --selector app=<application> --include-transitive-needs
```

> Fresh eyes note: change these to use ./bin/ck8s ops helmfile <sc|wc> instead?

Author: Tried it, worked just as well, with the added benefit that it doesn't take into account the global env `KUBECONFIG` 😄
